src/capture/capture_video_ffmpeg.cpp

```diff
@@ -0,0 +1,681 @@
+                                                 const uint32_t num_sample_frames,
+                                                 const int16_t* sample_frames)
+{
+	if (!audio_encoder.is_initalised) {
+		if (!audio_encoder.init(sample_rate)) {
+			LOG_ERR("FFmpeg: Failed to init audio encoder");
+			return;
+		}
+		bool init_muxer = false;
+		if (!muxer.is_initalised && video_encoder.is_initalised) {
+			init_muxer = muxer.init(video_encoder, audio_encoder);
+			if (!init_muxer) {
+				LOG_ERR("FFmpeg: Failed to init muxer");
+				audio_encoder.free();
+				return;
+			}
+		}
+		mutex.lock();
+		audio_encoder.is_initalised = true;
+		muxer.is_initalised         = muxer.is_initalised || init_muxer;
+		mutex.unlock();
+		waiter.notify_all();
+	} else if (audio_encoder.sample_rate != sample_rate) {
+		capture_video_finalise();
+		if (!audio_encoder.init(sample_rate)) {
+			LOG_ERR("FFmpeg: Failed to init audio encoder");
+			return;
+		}
+		mutex.lock();
+		audio_encoder.is_initalised = true;
+		mutex.unlock();
+		waiter.notify_all();
+	}
+
+	// Thought about writing a BulkEnqueue method that takes a pointer
+	// To avoid the temporary std::vector
+	// But this entire function only takes 5 microseconds average on my
+	// computer and removing the std::vector didn't help it that much
+	const size_t num_frames = num_sample_frames * 2;
+	std::vector<int16_t> audio_data(sample_frames, sample_frames + num_frames);
+	audio_encoder.queue.BulkEnqueue(audio_data, num_frames);
+}
+
+void FFmpegEncoder::stop_queues()
+{
+	std::unique_lock<std::mutex> lock(mutex);
+
+	// Set these first so none of the threads start another iteration
+	// Before the encoders get re-initalised
+	audio_encoder.is_initalised = false;
+	video_encoder.is_initalised = false;
+	muxer.is_initalised         = false;
+
+	audio_encoder.queue.Stop();
+	video_encoder.queue.Stop();
+	while (audio_encoder.is_working || video_encoder.is_working) {
+		waiter.wait(lock);
+	}
+	muxer.queue.Stop();
+	while (muxer.is_working) {
+		waiter.wait(lock);
+	}
+	assert(audio_encoder.queue.IsEmpty());
+	assert(video_encoder.queue.IsEmpty());
+	assert(muxer.queue.IsEmpty());
+	muxer.free();
+	audio_encoder.free();
+	video_encoder.free();
+}
+
+void FFmpegEncoder::start_queues()
+{
+	audio_encoder.queue.Start();
+	video_encoder.queue.Start();
+	muxer.queue.Start();
+}
+
+void FFmpegEncoder::capture_video_finalise()
+{
+	stop_queues();
+	start_queues();
+}
+
+bool FFmpegMuxer::init(FFmpegVideoEncoder& video_encoder,
+                       FFmpegAudioEncoder& audio_encoder)
+{
+	const int32_t output_file_index = get_next_capture_index(CaptureType::Video_MP4);
+	const std::string output_file_path = generate_capture_filename(CaptureType::Video_MP4,
+	                                                          output_file_index).string();
+
+	avformat_alloc_output_context2(&format_context,
+	                               nullptr,
+	                               nullptr,
```

Comment:
Can you extract these `nullptr`s to meaningful constants? What are we setting to `nullptr`?

Reply:
I'm happy to extract these if you like.  Just curious though, do you not use a language server or anything?  It's really helpful for things like this.  This is just clangd in vscode:

![clangd](https://github.com/dosbox-staging/dosbox-staging/assets/8128047/8063daf1-02cd-401d-b38f-6fae1f605e40)

Reply:
No, I just use plain old vim (NeoVim, technically). I like to keep it simple, plus there are no language servers when you're reading code reviews, diffs, and so on.

In languages that I use at work that have named args, we generally use named args for everything, it's super nice (e.g. in Kotlin).

Reply:
Extracted to constants